File: src/main/resources/META-INF/modules/app/lastPosts.js
define("app/lastPosts", ["require", "jquery", "app/templates"], function(require, $, templates) {
	function LastPosts(spec) {
		var _this = this;
		this.spec = spec;

		var source = require(["app/sources/" + this.spec.source], function(source) {
			var lastPosts = templates.lastPosts();
		
			var html = lastPosts(source);
			$('#' + _this.spec.id).html(html);
		});
	}
	
	function init(spec) {
		new LastPosts(spec);
	}

	return {
		init: init
	}
});

File: src/main/java/info/blogstack/components/LastPosts.java
package info.blogstack.components;

import info.blogstack.persistence.jooq.tables.records.SourceRecord;

import org.apache.tapestry5.annotations.Component;
import org.apache.tapestry5.annotations.Environmental;
import org.apache.tapestry5.annotations.Parameter;
import org.apache.tapestry5.annotations.Property;
import org.apache.tapestry5.corelib.components.Any;
import org.apache.tapestry5.json.JSONObject;
import org.apache.tapestry5.services.javascript.JavaScriptSupport;

public class LastPosts {

	@Parameter
	@Property
	private SourceRecord source;
	
	@Component
	private Any lastPosts;
	
	@Environmental
	private JavaScriptSupport support;
	
	void afterRender() {
		JSONObject spec = new JSONObject();
		spec.put("id", lastPosts.getClientId());
		spec.put("source", (source == null) ? "blogstack" : source.getAlias());
		spec.put("name", (source == null) ? "Blog Stack" : source.getName());
		support.require("app/lastPosts").invoke("init").with(spec);
	}
}